src/framework/theme/components/dialog/dialog.service.ts

```diff
@@ -0,0 +1,222 @@
+/**
+ * @license
+ * Copyright Akveo. All Rights Reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ */
+
+import { Inject, Injectable, Injector, TemplateRef, Type } from '@angular/core';
+import { fromEvent as observableFromEvent } from 'rxjs';
+import { filter } from 'rxjs/operators';
+
+import {
+  NbComponentPortal,
+  NbGlobalPositionStrategy,
+  NbOverlayRef,
+  NbOverlayService,
+  NbPortalInjector,
+  NbPositionBuilderService,
+  NbScrollStrategy,
+  NbTemplatePortal,
+} from '../cdk';
+import { NB_DOCUMENT } from '../../theme.options';
+import { NB_DIALOG_CONFIG, NbDialogConfig } from './dialog-config';
+import { NbDialogRef } from './dialog-ref';
+import { NbDialogContainerComponent } from './dialog-container';
+
+
+/**
```

Comment:
GRAMMARLY!
```
The `NbDialogService` service provides a functionality to open a dialog.

@stacked-example(Showcase, dialog/dialog-showcase.component)

A new dialog is opened by calling the `open` method with a component to be loaded and an optional configuration.
`open` method will return `NbDialogRef` that can be used for the further manipulations with the dialog.

```ts
const dialogRef = this.dialogService.open(MyDialogComponent, { ... });
```

`NbDialogRef` gives a capability to access a reference to the rendered dialog component,
destroy dialog and some other options described below.

Also, you can inject `NbDialogRef` to a host component constructor:

```ts
this.dialogService.open(MyDialogComponent, { ... });

// my-dialog.component.ts
constructor(protected dialogRef: NbDialogRef) {
}

close() {
  this.dialogRef.close();
}
```

Instead of a component, you can create the dialog by passing a `TemplateRef`:

@stacked-example(Template ref, dialog/dialog-template.component)

The dialog may return the result through `NbDialogRef`. Calling component can receive this result with the `onClose`
stream of `NbDialogRef`.

@stacked-example(Result, dialog/dialog-result.component)

### Configuration

As we mentioned above, `open` method of the `NbDialogService` may receive optional configuration options.
Also, you can provide a global dialogs configuration through the `NbDialogModule.forRoot({ ... })` method.

This config may contain the following options:

`context` - both, template and component may receive data through `config.context` property.
For components, this data will be assigned through inputs.
For templates, you can access it inside of a template as $implicit.

```ts
this.dialogService.open(template, { context: 'pass data in template' });
```

```html
<ng-template let-some-additional-data>
  {{ some-additional-data }}
<ng-template/>
```

`hasBackdrop` - determines whether the dialog will be having a backdrop background, `true` by default.
@stacked-example(Backdrop, dialog/dialog-has-backdrop.component)

`closeOnBackdropClick` - close dialog on backdrop click if true, `true` by default.
@stacked-example(Backdrop click, dialog/dialog-backdrop-click.component)

`closeOnEsc` - close dialog on Escape button on the keyboard, `true` by default.
@stacked-example(Escape hit, dialog/dialog-esc.component)

`hasScroll` - Disables scroll on content under dialog if true and does nothing otherwise, `false` by default.
You can open the example in a separate window and try to scroll the page:
@stacked-example(Scroll, dialog/dialog-scroll.component)

`autoFocus` - Focuses dialog automatically after open if `true`. It's useful to prevent misclicks on
trigger elements and opening multiple dialogs, `true` by default.

As you can see, if you open dialog with `autoFocus` on, the dialog will focus first focusable element
or just blur previously focused elements automatically.
Otherwise, without `autoFocus`, the focus will stay on the previously focused element.
You can open the example in a separate window and try to click on the button without focus:
and then hit space any times. Multiple same dialogs will be opened.
@stacked-example(Auto focus, dialog/dialog-auto-focus.component)
```